src/utils/io/progress_log.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use humantime::format_duration;
+use std::time::{Duration, Instant};
+
+use parking_lot::Mutex;
+use pin_project_lite::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::io::ReadBuf;
+
+use log::info;
+
+const UPDATE_FREQUENCY: Duration = Duration::from_millis(5000);
+
+pin_project! {
+    /// Wraps an iterator to display its progress.
+    pub struct WithProgressStream<S> {
+        #[pin]
+        stream: S,
+        progress: WithProgress,
+    }
+}
+
+impl<R: tokio::io::AsyncRead> tokio::io::AsyncRead for WithProgressStream<R> {
```

Comment:
This is a misnomer: `WithProgressStream` is too specific a name - note it's implementing `AsyncRead`, not `Stream`, which is suprising. See previous comment above also.

Reply:
Any suggestion for a better name?

Reply:
`WithProgressAsync` ?

Reply:
I think we should just have a top-level `WithProgress`. Implementations of sync traits shouldn't care about the pinned inner. 

If we can't do that right now, then we should keep this as `WithProgress`, and rename `ProgressLog` to something like `WithProgressRaw`, with a view to hiding it.

It's an iterator|reader|writer|stream _with progress_ logs